src/Debug/redux-devtool.ts

```diff
@@ -0,0 +1,333 @@
+  return ([action, model]) => (action.type === 'MESSAGE' ? connection.send(action.payload, model) : undefined)
+}
+
+/**
+ * Parses a `JUMP` message.
+ *
+ * @since 0.5.0
+ */
+function parseJump<Model>(msg: Monitor): Either<string, Model> {
+  return pipe(
+    E.parseJSON(msg.state, E.toError),
+    E.bimap(e => e.message, u => u as Model)
+  )
+}
+
+/**
+ * Parses a `ROLLBACK` message.
+ *
+ * @since 0.5.0
+ */
+function parseRollback<Model>(msg: Monitor): Either<string, Model> {
+  return pipe(
+    E.parseJSON(msg.state, E.toError),
+    E.bimap(e => e.message, u => u as Model)
+  )
+}
+
+/**
+ * Parses an `IMPORT_STATE` message.
+ *
+ * @since 0.5.0
+ */
+function parseImportState<Model>(msg: Monitor): Either<string, LiftedState<Model>> {
+  return typeof msg.payload.nextLiftedState === 'object' && msg.payload.nextLiftedState !== null
+    ? E.right(msg.payload.nextLiftedState as LiftedState<Model>)
+    : E.left('IMPORT_STATE message has some bad payload...')
+}
+
+/**
+ * Parses a `TOGGLE_ACTION` message.
+ *
+ * @since 0.5.0
+ */
+function parseToggleAction<Model>(msg: Monitor): Either<string, [number, LiftedState<Model>]> {
+  const getId = pipe(
+    msg.payload.id,
+    E.fromNullable('TOGGLE_ACTION message has some bad payload...'),
+    E.map<any, number>(x => x)
+  )
+
+  const parseState = pipe(
+    E.parseJSON(msg.state, E.toError),
+    E.bimap(e => e.message, u => u as LiftedState<Model>)
+  )
+
+  return sequenceTEither(getId, parseState)
+}
+
+/**
+ * Handles toggling a specific action (identified by `id` parameter).
+ *
+ * It re-executes all the actions (and updates the store) excluding the toggled one.
+ *
+ * The implementation is taken from [MobX dev tool integration](https://github.com/zalmoxisus/mobx-remotedev/blob/master/src/monitorActions.js#L22)
+ *
+ * @since 0.5.0
+ */
+function toggleAction<Model, Msg>(
+  dispatch: Dispatch<MsgWithDebug<Model, Msg>>
+): (current: Model, id: number, liftedState: LiftedState<Model>) => IO<LiftedState<Model>> {
+  return (current, id, liftedState) => () => {
+    const state = JSON.parse(JSON.stringify(liftedState)) // poor man deep clone...
+
+    const { skippedActionIds, stagedActionIds, computedStates, actionsById } = state
+
+    const skippedIndex = skippedActionIds.indexOf(id)
+    const skipped = skippedIndex !== -1
+    const actionIndex = stagedActionIds.indexOf(id)
+
+    if (actionIndex === -1) {
+      return state
+    }
+
+    dispatch({ type: '__DebugUpdateModel__', payload: computedStates[actionIndex - 1].state })
+
+    const start = skipped ? actionIndex : actionIndex + 1
+    const end = stagedActionIds.length
+
+    for (let i = start; i < end; i++) {
+      const currentActionId = stagedActionIds[i]
+
+      if (i !== actionIndex && skippedActionIds.indexOf(currentActionId) !== -1) {
+        continue // it's already skipped
+      }
+
+      dispatch(actionsById[currentActionId].action as Msg)
```

Comment:
this will run all Cmds that are produced by this Msg, which can produce new undesirable Msgs during time travelling (see gif below). The only way I am aware of to fix this, (since Task cannot be interpreted in any other way than being run or being ignored) is to somehow ignore these Cmds.
![debug](https://user-images.githubusercontent.com/1811622/69420209-8634ab00-0d1e-11ea-81de-fbd767719153.gif)

Reply:
More flexible approach could be to give the user information about debuging status, with something like this (sketching only, don't take names too seriously):
```js
export function programWithDebuggerMoreFlexible<Model, Msg, Dom>(
  init: (isTimeTravelling: boolean) => [Model, cmd.Cmd<Msg>],
  update: (isTimeTravelling: boolean) => (msg: Msg, model: Model) => [Model, cmd.Cmd<Msg>],
  view: (model: Model) => Html<Dom, Msg>,
  subscriptions?: (isTimeTravelling: boolean) => (model: Model) => Sub<Msg>
): Program<Model, MsgWithDebug<Model, Msg>, Dom> {
  //...
}

// with default being
export function programWithDebugger<Model, Msg, Dom>(
  init: [Model, cmd.Cmd<Msg>],
  update: (msg: Msg, model: Model) => [Model, cmd.Cmd<Msg>],
  view: (model: Model) => Html<Dom, Msg>,
  subscriptions: (model: Model) => Sub<Msg> = () => none
): Program<Model, MsgWithDebug<Model, Msg>, Dom> {
  return programWithDebuggerMoreFlexible(
    _ => [init[0], Cmd.none],
    _ => (msg: Msg, model: Model) => {
      const [model1, _] = update(msg, model)
      return [model1, Cmd.none]
    },
    view,
    subscriptions && _ => Sub.none
  )
}
```
This way user could choose different strategy form Cmds in time travelling mode

Reply:
Thank you very much Milos (I owe you for the hints in your previous reviews :wink:)

I would like to avoid requesting other actions to the user: in my vision the debugging tool usage should be as transparent as possible (only a single line like `const program = process.env.NODE_ENV === 'production' ? ....`)

Furthermor, how should be handled that `isTimeTravelling` flag? The user should inject in the program's source some specific code for the redux devtool? 

I guess that time travelling can be sloppy in many standard redux application with async actions (and effects)

Also, I have to confess that I hugely "take inspiration"  (:smile:) for the `TOGGLE_ACTION` handler - which controls the skip and time travelling features - from the official [MobX integration](https://github.com/zalmoxisus/mobx-remotedev/blob/master/src/monitorActions.js#L22), but I barely understand how it works... :grimacing:

Reply:
@minedeljkovic you're right: the extension works correctly with a standard redux app with async actions (`redux-thunk`); [check this codesandbox](https://codesandbox.io/s/hardcore-hill-msds0)

It seems that implementation of the `TOGGLE_ACTION` handler is quite different from the MobX one (it is probably hidden somewhere [here](https://github.com/reduxjs/redux-devtools/blob/master/packages/redux-devtools-instrument/src/instrument.js))

I'm quite stuck on this... Maybe I should re-use "as-is" the implementation in your gist...

Reply:
> I guess that time travelling can be sloppy in many standard redux application with async actions (and effects)

True 😕 , but it can also greatly simplify and speed up QA and customer support process 👍 . In company I work for it became really important part of the process.

These steps are one way to drop cmds comming from time travelling (to reach parity with redux-thunk example):

1. Introduce one more special msg for time travelling:
```js
export type MsgWithDebug<Model, Msg> = Msg | { type: '__DebugUpdateModel__'; payload: Model } | { type: '__DebugApplyMsg__'; payload: Msg }

```
here
https://github.com/gcanti/elm-ts/blob/77e1f5e44dd4de6ec71d536ca4ff460969e9728c/src/Debug/commons.ts#L48

2. Handle this special msg like:
```js
    if ('type' in msg) {
      switch (msg.type) {
        case '__DebugUpdateModel__':
          return [msg.payload, cmd.none]
        case '__DebugApplyMsg__': {
          const [newModel/*, drop cmd*/] = update(msg.payload, model)
          return [newModel, cmd.none]
        }
      }
    }
```
here
https://github.com/gcanti/elm-ts/blob/77e1f5e44dd4de6ec71d536ca4ff460969e9728c/src/Debug/index.ts#L69-L71

3. Dispatch this special msg:
```js
dispatch({ type: '__DebugApplyMsg__', payload: actionsById[currentActionId].action as Msg })
```
here
https://github.com/gcanti/elm-ts/blob/77e1f5e44dd4de6ec71d536ca4ff460969e9728c/src/Debug/redux-devtool.ts#L320

Reply:
It looks promising. I'll try right now

Reply:
@minedeljkovic it seems working! Please, check it out

Reply:
This is working, now! 👍